File: go-service/messaging/publisher.go
package messaging

import (
	"encoding/json"
	"log/slog"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

// SensorEvent represents the sensor.updated event published to RabbitMQ.
type SensorEvent struct {
	Event     string  `json:"event"`
	SensorID  string  `json:"sensor_id"`
	Value     float64 `json:"value"`
	Type      string  `json:"type"`
	Unit      string  `json:"unit"`
	Timestamp string  `json:"timestamp"`
	TraceID   string  `json:"trace_id"`
}

// EventPublisher publishes sensor events to RabbitMQ.
// Uses a fanout exchange so all bound queues (e.g. alert services) receive every event.
// Reconnects automatically on failure — publish errors are logged and swallowed
// so a RabbitMQ outage never breaks the HTTP response.
type EventPublisher struct {
	url     string
	conn    *amqp.Connection
	channel *amqp.Channel
}

// NewEventPublisher creates a new publisher. It attempts an initial connection
// but does not fail if RabbitMQ is not yet available.
func NewEventPublisher(url string) *EventPublisher {
	p := &EventPublisher{url: url}
	if err := p.connect(); err != nil {
		slog.Warn("RabbitMQ not available at startup — will retry on first publish", "error", err)
	}
	return p
}

func (p *EventPublisher) connect() error {
	conn, err := amqp.Dial(p.url)
	if err != nil {
		return err
	}

	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return err
	}

	if err := ch.ExchangeDeclare("sensor_events", "fanout", true, false, false, false, nil); err != nil {
		ch.Close()
		conn.Close()
		return err
	}

	p.conn = conn
	p.channel = ch
	slog.Info("Connected to RabbitMQ for publishing")
	return nil
}

// PublishSensorUpdated publishes a sensor.updated event to the sensor_events exchange.
// If RabbitMQ is unavailable the error is logged and the call returns silently.
func (p *EventPublisher) PublishSensorUpdated(sensorID string, value float64, sensorType, unit, traceID string) {
	event := SensorEvent{
		Event:     "sensor.updated",
		SensorID:  sensorID,
		Value:     value,
		Type:      sensorType,
		Unit:      unit,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		TraceID:   traceID,
	}

	body, err := json.Marshal(event)
	if err != nil {
		slog.Error("Failed to marshal sensor event", "error", err)
		return
	}

	if err := p.publish(body); err != nil {
		slog.Warn("Failed to publish sensor event, attempting reconnect", "error", err)
		p.conn = nil
		p.channel = nil

		if reconnErr := p.connect(); reconnErr != nil {
			slog.Warn("RabbitMQ reconnect failed — event dropped", "sensor_id", sensorID, "error", reconnErr)
			return
		}

		if err := p.publish(body); err != nil {
			slog.Warn("Publish failed after reconnect — event dropped", "sensor_id", sensorID, "error", err)
			return
		}
	}

	slog.Info("Published sensor.updated event", "sensor_id", sensorID, "value", value, "trace_id", traceID)
}

func (p *EventPublisher) publish(body []byte) error {
	if p.channel == nil {
		return amqp.ErrClosed
	}
	return p.channel.Publish(
		"sensor_events", // exchange
		"",              // routing key (fanout ignores it)
		false,           // mandatory
		false,           // immediate
		amqp.Publishing{
			DeliveryMode: amqp.Persistent,
			ContentType:  "application/json",
			Body:         body,
		},
	)
}

File: python-service/messaging/publisher.py
"""RabbitMQ event publisher for sensor update events."""

import json
import logging
from datetime import datetime, timezone

import pika

logger = logging.getLogger("sensor_service")


class EventPublisher:
    """Publishes sensor events to RabbitMQ.

    Uses a fanout exchange so all bound queues (e.g. alert services)
    receive every event. Reconnects automatically on failure.
    """

    def __init__(self, rabbitmq_url: str):
        self.rabbitmq_url = rabbitmq_url
        self._connection = None
        self._channel = None

    def connect(self):
        """Establish connection to RabbitMQ and declare the exchange.

        Tolerates failure — logs a warning and leaves the connection as None
        so publish_sensor_updated can retry on the next call.
        """
        try:
            params = pika.URLParameters(self.rabbitmq_url)
            self._connection = pika.BlockingConnection(params)
            self._channel = self._connection.channel()
            self._channel.exchange_declare(
                exchange="sensor_events", exchange_type="fanout", durable=True
            )
            logger.info("Connected to RabbitMQ for publishing")
        except Exception as e:
            logger.warning(
                "RabbitMQ not available at startup — will retry on first publish: %s", str(e)
            )
            self._connection = None
            self._channel = None

    def publish_sensor_updated(
        self, sensor_id: str, value: float, sensor_type: str, unit: str, trace_id: str
    ) -> None:
        """Publish a sensor.updated event to the sensor_events exchange.

        Args:
            sensor_id: ID of the updated sensor.
            value: New sensor value.
            sensor_type: Type of sensor (temperature, humidity, etc.).
            unit: Unit of measurement.
            trace_id: UUID generated at the HTTP handler to trace this event end-to-end.
        """
        event = {
            "event": "sensor.updated",
            "sensor_id": sensor_id,
            "value": value,
            "type": sensor_type,
            "unit": unit,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "trace_id": trace_id,
        }

        try:
            if not self._channel or self._channel.is_closed:
                self.connect()

            self._channel.basic_publish(
                exchange="sensor_events",
                routing_key="",
                body=json.dumps(event),
                properties=pika.BasicProperties(delivery_mode=2),  # persistent
            )
            logger.info(
                "Published sensor.updated event",
                extra={"sensor_id": sensor_id, "value": value, "trace_id": trace_id},
            )
        except Exception as e:
            logger.warning("Failed to publish event: %s", str(e))
            self._connection = None
            self._channel = None
